genrl/deep/agents/dqn/base.py

```diff
@@ -0,0 +1,164 @@
+        self.max_epsilon = max_epsilon
+        self.min_epsilon = min_epsilon
+        self.epsilon_decay = epsilon_decay
+        self.layers = layers
+        self.lr = lr
+
+        self.empty_logs()
+        self.create_model()
+
+    def create_model(self) -> None:
+        input_dim, action_dim, _, _ = get_env_properties(self.env, self.network_type)
+
+        self.model = get_model("v", self.network_type)(
+            input_dim, action_dim, "Qs", self.layers
+        )
+        self.target_model = deepcopy(self.model)
+
+        self.replay_buffer = PushReplayBuffer(self.replay_size)
+        self.optimizer = opt.Adam(self.model.parameters(), lr=self.lr)
+
+    def update_target_model(self) -> None:
+        self.target_model.load_state_dict(self.model.state_dict())
+
+    def update_params_before_select_action(self, timestep: int) -> None:
+        self.timestep = timestep
+        self.epsilon = self.calculate_epsilon_by_frame()
+
+    def select_action(
+        self, state: np.ndarray, deterministic: bool = False
+    ) -> np.ndarray:
+
+        if not deterministic:
+            if np.random.rand() < self.epsilon:
+                return np.asarray(self.env.sample())
+
+        state = torch.FloatTensor(state)
+        q_values = self.model(state).detach().numpy()
+
+        return np.argmax(q_values, axis=-1)
+
+    def get_q_values(self, states, actions) -> torch.Tensor:
+        q_values = self.model(states).gather(1, actions)
+        return q_values
+
+    def get_target_q_values(self, next_states, rewards, dones):
+        next_q_target_values = self.target_model(next_states)
+        max_next_q_target_values = next_q_target_values.max(1)[0]
+        target_q_values = rewards + self.gamma * torch.mul(
+            max_next_q_target_values, (1 - dones)
+        )
+        return target_q_values.unsqueeze(-1)
+
+    def update_params(self, update_interval: int) -> None:
+        for timestep in range(update_interval):
+            states, actions, rewards, next_states, dones = self.replay_buffer.sample(
+                self.batch_size
+            )
+
+            states = states.reshape(-1, *self.env.obs_shape)
+            actions = actions.reshape(-1, *self.env.action_shape).long()
+            next_states = next_states.reshape(-1, *self.env.obs_shape)
+
+            rewards = torch.FloatTensor(rewards)
+            dones = torch.FloatTensor(dones)
+
+            q_values = self.get_q_values(states, actions)
+            target_q_values = self.get_target_q_values(next_states, rewards, dones)
+            # print(q_values.shape)
+            # print(target_q_values.shape)
+            # print("\n\n")
+
+            loss = F.mse_loss(q_values, target_q_values)
+            self.logs["value_loss"].append(loss.item())
+
+            self.optimizer.zero_grad()
+            loss.backward()
+            self.optimizer.step()
+
+            if timestep % update_interval == 0:
+                self.update_target_model()
```

Comment:
These steps would be same in all DQNs? So we could put it in a different function altogether.
Would also reduce complexity

Reply:
Yeah. I changed this locally.